common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionHandler.java

```diff
@@ -51,7 +67,38 @@ public void onDeviceTelemetry(MqttPublishMessage mqttMsg) throws AdaptorExceptio
 
     @Override
     protected GatewayDeviceSessionContext newDeviceSessionCtx(GetOrCreateDeviceFromGatewayResponse msg) {
-         return new GatewayDeviceSessionContext(this, msg.getDeviceInfo(), msg.getDeviceProfile(), mqttQoSMap, transportService);
+        return new GatewayDeviceSessionContext(this, msg.getDeviceInfo(), msg.getDeviceProfile(), mqttQoSMap, transportService);
+    }
+
+    public void onGatewayDisconnect() {
+        this.onDevicesDisconnect();
+        latencyService.onDeviceDisconnect(gateway.getDeviceId());
+    }
+
+    public void onGatewayUpdate(TransportProtos.SessionInfoProto sessionInfo, Device device, Optional<DeviceProfile> deviceProfileOpt) {
+        this.onDeviceUpdate(sessionInfo, device, deviceProfileOpt);
+        latencyService.onDeviceUpdate(sessionInfo, gateway.getDeviceId());
+    }
+
+    public void onGatewayDelete(DeviceId deviceId) {
+        latencyService.onDeviceDelete(deviceId);
+    }
+
+    public void onGatewayLatency(MqttPublishMessage mqttMsg) throws AdaptorException {
+        int msgId = getMsgId(mqttMsg);
+        ByteBuf payloadData = mqttMsg.payload();
+        String payload = payloadData.toString(UTF8);
+        if (payload == null) {
+            log.debug("[{}][{}][{}] Payload is empty!", gateway.getTenantId(), gateway.getDeviceId(), sessionId);
+            throw new AdaptorException(new IllegalArgumentException("Payload is empty!"));
+        }
+        long ts = System.currentTimeMillis();
+        try {
+            latencyService.process(deviceSessionCtx.getSessionInfo(), gateway.getDeviceId(), JacksonUtil.fromString(payload, new TypeReference<>() {}), ts);
```

Comment:
Probably it will be better to use a default behavior, with conversion in adapters, don’t you think so?